### PR TITLE
[WIP]: API ExtensionDtype for DTA & TDA

### LINKED ISSUE #24662: dtype for tz-naive DatetimeARray and TimedeltaArray
Right now, `Datetime.Array.dtype` can be either `np.dtype('M8[ns])` or a `DatetimeTZDtype`, depending on whether the values are tz-naive or tz-aware. This means that while `DatetimeArray[tz-naive]` is an *instance* of `ExtensionArray`, it doesn't satisfy the minimum ExtensionArray API, which requires that `array.dtype` be an `ExtensionDtype`.

```python
In [4]: pd.date_range('2000', periods=4)._data.dtype
Out[4]: dtype('<M8[ns]')

In [5]: pd.date_range('2000', periods=4, tz="CET")._data.dtype
Out[5]: datetime64[ns, CET]
```

The causes some type-unsoundness for places that are supposed to return an ExtensionArray. The two most prominent being `pd.array` and `Series.array`. As an example, following isn't necessarily safe code

```python
def f(ser: Series) -> Callable: 
    return ser.array.dtype.construct_array_type
```

that will fail for tz-naive datetime data, because its `.dtype` is a NumPy dtype.

---

Proposal:

1. Make a `DatetimeDtype` (or allow `DatetimeTZDtype` to have `tz=None`). Make a `TimedeltaArray.dtype`
2. Ensure that `DatetimeArray.dtype` and `TimedeltaArray.dtype` is always an ExtensionDtype
3. Wrap `Series.dtype` and `DatetimeIndex.dtype` to *continue* returning the NumPy dtype

The last step is to avoid breaking code relying on `Series[tz-naive].dtype` being a NumPy dtype.

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/24662

TODO:

- [ ] Dedicated tests for new is_<foo> types
- [ ] Run the extension dtype tests for new dtypes
- [ ] (maybe?) see if the new EA tests pass for the new arrays. Tricky, since they aren't allowed internally. g
- [ ] failing dtype validation tests

This is going to fail. Going offline for a bit, but putting this up right now.

I'm going to clean up the changes in `internals` and the series constructor.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Still a WIP, but probably not going to have time to update this any further through the end of the week.

File: pandas/core/dtypes/dtypes.py [hunk 1]
```diff
@@ -573,25 +573,112 @@ def _is_boolean(self):
...
+
+        Returns
+        -------
+        type
+        """
+        from pandas.core.arrays import DatetimeArray
+        return DatetimeArray
+
+    @classmethod
+    def construct_from_string(cls, string):
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py]: I think you need to move this to the sub-class as DatetimeDtype would actually accept this but raise on a construction error (as you can't pass a tz). I would maybe instead just have a unified constructor that accepts ``tz=None`` for naive.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
> I would maybe instead just have a unified constructor that accepts tz=None for naive.

@jreback do you also mean just having a single class internally? `DatetimeDtype(tz=None)` is the naive one and `DatetimeDtype(tz=tz)` is the aware one (and `DatetimeTzDtype` can be kept as an alias for now for backwards compatibility) 

I didn't really think about any possible technical implications of that, but it seems logical to me to do that, since we also have a single DatetimeArray.

**Comment by TomAugspurger** [comment]
Sorry I put this up in haste and didn't explain things well. Initially, I tried that, but ran into some difficulties. It turns out that having `pandas_dtype("M8[ns]")` return a non NumPy dtype breaks a lot of stuff. So I split them apart, and don't register the `DatetimeDtype`.

I'm not sure if it's possible to both

1. Have a DatetimeDtype that registers datetime64[ns, tz], but not datetime64[ns]
2. Have construct_from_string always work correctly

but I can try (later, not today).

**Comment by jreback** [comment]
yeah I think this is a great idea @TomAugspurger but also possibly to involve changing some tricky code; would prefer to push this to 0.25

**Comment by TomAugspurger** [comment]
The main downside with pushing to 0.25 is it'll be an API breaking change for `DatetimeArray.dtype`, and `TimedeltaArray.dtype` with no path for deprecation.

---

just pushed an update, but still a WIP.

**Comment by TomAugspurger** [comment]
Somewhat awkward API corner: what's the expected types for

1. `DatetimeArray.astype("M8[ns]")`
2. `DatetimeArray.astype(np.dtype("M8[ns]"))`
3. `DatetimeArray.astype(pd.DatetimeDtype())`

Previously, we'd use `.astype('M8[ns]')` to convert to an ndarray. But now that we have a dtype roughly equivalent to `M8[ns]`, should we keep doing that?

Another one, what should we do about `pd.DatetimeDtype() == "M8[ns]"` or `== np.dtpye("M8[ns]")`.

NumPy doesn't make this easy, since they would raise if the operands were flipped, rather than returning NotImplemented.

**Comment by jorisvandenbossche** [comment]
For the expected types, I also have been wondering about this in general. So eg also for `DTA.astype("int64")`. 
Because we now have `PandasArray` and `PandasDtype`, so we could also return those? Similarly as `pd.array(DTA, dtype='int64')` does.

**Comment by TomAugspurger** [comment]
Yeah, it's not obvious to me what's best here.

On the one hand, you could say that `.astype(np.dtype(.))` should always return an ndarray.

On the other hand, is it weird for `.astype(np.dtype("int64"))` to not match `.astype("int64")`? I suppose we'll need to sort all this out.

I'm really not sure what to do here :/

**Comment by h-vetinari** [comment]
> @TomAugspurger: Yeah, it's not obvious to me what's best here.
On the one hand, you could say that `.astype(np.dtype(.))` should always return an ndarray.
On the other hand, is it weird for `.astype(np.dtype("int64"))` to not match `.astype("int64")`? I suppose we'll need to sort all this out.
I'm really not sure what to do here :/

How about `.astype` and `pd.array` always returning `PandasDType` (yes, even for an `np.dtype`)? If the user really wants an `ndarray`, they can then call `to_numpy`. That would in many ways be more consistent - and could later be extended to the point that the *only* way to get an `ndarray` from pandas is to call `to_numpy`.

**Comment by jbrockmendel** [comment]
This has been inactive for a while.  Still a goal?

**Comment by TomAugspurger** [comment]
I'm not sure. The API changes are tricky to get right.

Regardless, I don't think things will be done on this branch. I'll post an update on the issue.

**Pull request closed without merging**